rust/src/metrics/mod.rs

```diff
@@ -0,0 +1,350 @@
+        use super::*;
+
+        pub fn update(total_number: u64) -> WriteQuery {
+            RoundTotalNumber {
+                time: Timestamp::Now.into(),
+                round_total_number: total_number,
+            }
+            .into_query(Measurement::StateMachine.to_string())
+        }
+    }
+
+    pub mod successful {
+        use super::*;
+
+        pub fn increment() -> WriteQuery {
+            RoundSuccessful {
+                time: Timestamp::Now.into(),
+                round_successful: 1,
+            }
+            .into_query(Measurement::StateMachine.to_string())
+        }
+    }
+}
+
+pub mod message {
+    use super::models::{
+        Measurement,
+        MessageDiscarded,
+        MessageRejected,
+        MessageSum,
+        MessageSum2,
+        MessageUpdate,
+    };
+    use crate::state_machine::phases::PhaseName;
+    use influxdb::{InfluxDbWriteable, Timestamp, WriteQuery};
+    pub mod sum {
+        use super::*;
+
+        pub fn increment(round_id: u64, phase: PhaseName) -> WriteQuery {
+            MessageSum {
+                time: Timestamp::Now.into(),
+                message_sum: 1,
+                round_id,
+                phase: phase as u8,
+            }
+            .into_query(Measurement::StateMachine.to_string())
+        }
+    }
+
+    pub mod update {
+        use super::*;
+
+        pub fn increment(round_id: u64, phase: PhaseName) -> WriteQuery {
+            MessageUpdate {
+                time: Timestamp::Now.into(),
+                message_update: 1,
+                round_id,
+                phase: phase as u8,
+            }
+            .into_query(Measurement::StateMachine.to_string())
+        }
+    }
+
+    pub mod sum2 {
+        use super::*;
+
+        pub fn increment(round_id: u64, phase: PhaseName) -> WriteQuery {
+            MessageSum2 {
+                time: Timestamp::Now.into(),
+                message_sum2: 1,
+                round_id,
+                phase: phase as u8,
+            }
+            .into_query(Measurement::StateMachine.to_string())
+        }
+    }
+
+    pub mod discarded {
+        use super::*;
+
+        pub fn increment(round_id: u64, phase: PhaseName) -> WriteQuery {
+            MessageDiscarded {
+                time: Timestamp::Now.into(),
+                message_discarded: 1,
+                round_id,
+                phase: phase as u8,
+            }
+            .into_query(Measurement::StateMachine.to_string())
+        }
+    }
+
+    pub mod rejected {
+        use super::*;
+
+        pub fn increment(round_id: u64, phase: PhaseName) -> WriteQuery {
+            MessageRejected {
+                time: Timestamp::Now.into(),
+                message_rejected: 1,
+                round_id,
+                phase: phase as u8,
+            }
+            .into_query(Measurement::StateMachine.to_string())
+        }
+    }
+}
+
+pub async fn run_metric_service(mut metics_service: MetricsService) {
```

Comment:
One downside of this implementation is that we process the request one by one. If the queries are slow, the queries will accumulate in the channel. Maybe we could make that channel bounded to apply some backpressure. On the longer run, we could use tower to benefit from timeouts, backpressure and concurrent requests.